File: models/simulation_studies/latent_recovery/compute_statistics.R
# Use the MVN package to test whether the sample points are from a MVN distribution 

#!/usr/bin/env Rscript
# USAGE: Rscript compute_statistics.R X_Sample_Covariance.csv hyperparameters.yaml

library("MVN")

# Check if the correct number of arguments is provided
if (length(commandArgs(trailingOnly = TRUE)) < 2) {
    stop("USAGE: Rscript compute_statistics.R X_Sample_Covariance.csv")
}

# Get the config file path from the second argument
X_sample_cov_file <- commandArgs(trailingOnly = TRUE)[1]
config_file <- commandArgs(trailingOnly = TRUE)[2]

# Read the YAML file
yaml_data <- yaml::read_yaml(config_file)

# Access individual variables and assign them to R variables
SEED <- yaml_data$SEED
N <- yaml_data$N1
T <- yaml_data$T
Q <- yaml_data$Q
B <- yaml_data$B
d <- yaml_data$d
sigma <- yaml_data$sigma
N_replicas <- yaml_data$N_replicas
H <- yaml_data$H
spectral_radius <- yaml_data$spectral_radius
n_iter <- yaml_data$n_iter


df <- read.csv(X_sample_cov_file,header=FALSE)

component_to_visualise = 21

subset_indices <- component_to_visualise + seq(0,(N_replicas-1)*N,by=N)

df_component = df[subset_indices,]
# df_component = df[150:300,]

result = mvn(data = df_component,
            mvnTest = "hz",
            scale = TRUE,
            univariateTest = "AD",
            univariatePlot = "histogram",
            multivariatePlot = "qq",
            multivariateOutlierMethod = "none",
            showOutliers = FALSE, 
            showNewData = FALSE)

multi_norm <- result$multivariateNormality
print(multi_norm) 

uni_normality <- result$univariateNormality
print(uni_normality)

descriptives <- result$Descriptives
print(descriptives)

# save multi_norm to csv file for all components 
p_vals <- matrix(nrow = N, ncol = 1)
for (i in 1:N) {
    component_to_visualise <- i

    subset_indices <- component_to_visualise + seq(0, (N_replicas-1)*N, by = N)

    df_component <- df[subset_indices,]

    result <- mvn(data = df_component,
                mvnTest = "hz",
                scale = TRUE,
                univariateTest = "AD",
                univariatePlot = "histogram",
                multivariatePlot = "qq",
                multivariateOutlierMethod = "none",
                showOutliers = FALSE, 
                showNewData = FALSE)
    
    multi_norm <- result$multivariateNormality
    p_vals[i] <- multi_norm[["p value"]]
    
}

write.table(p_vals, file = "hz_pvals.csv", sep = ",", row.names = FALSE, col.names = FALSE, quote = FALSE)

File: models/simulation_studies/latent_recovery/compute_statistics.py
# Test whether the sample points are from a MVN distribution 

#!/usr/bin/env python3
# USAGE: ./visualise.py hyperparameters.yaml  X_Sample_Covariance.csv 

import numpy as np
from numpy.random import default_rng
import sys
import yaml
from numpy import genfromtxt
import plotly.graph_objects as go
import time
from scipy.stats import norm
import pingouin as pg 


with open(sys.argv[1], "r") as f:
    config = yaml.load(f, Loader=yaml.FullLoader)

###### CONFIG PARAMETERS ###### 
N = config['N1']
T = config['T']
Q = config['Q']
B = config['B']
d = config['d']
sigma = config['sigma']
SEED = config['SEED']
N_replicas = config['N_replicas']
H = config['H'] 
spectral_radius = config['spectral_radius']
n_iter = config['n_iter']

random_state = default_rng(seed=SEED)

###### READ IN DATA ###### 
X_S = genfromtxt(sys.argv[2],delimiter=",")
X_S = np.reshape(X_S,(N_replicas,N,d)) 

component_to_visualise = 4

X_S_contour_data = X_S[:,component_to_visualise,:]

hz_results = pg.multivariate_normality(X_S_contour_data, alpha=.05)

print(hz_results)

